internals/overlord/logstate/syslog/syslog.go

```diff
@@ -0,0 +1,300 @@
+
+const (
+	defaultMaxRequestEntries    = 100
+	defaultSyslogInitialBackoff = 100 * time.Millisecond
+	defaultSyslogMaxBackoff     = 10 * time.Second
+	defaultDialTimeout          = 10 * time.Second
+	canonicalPrivEnterpriseNum  = 28978
+)
+
+// Syslog Priority values - see RFC 5424 6.2.1
+const (
+	FacilityUserLevelMessage = 1
+	SeverityInformational    = 6
+)
+
+type ClientOptions struct {
+	MaxRequestEntries    int
+	TargetName           string
+	Location             string
+	SyslogInitialBackoff time.Duration
+	SyslogMaxBackoff     time.Duration
+	DialTimeout          time.Duration
+}
+
+type entryWithService struct {
+	service   string
+	Priority  int
+	Version   int
+	Timestamp string
+	PID       string
+	MsgID     string
+	Message   string
+}
+
+type Client struct {
+	options *ClientOptions
+
+	// To store log entries, keep a buffer of size 2*MaxRequestEntries with a
+	// sliding window "entries" of size MaxRequestEntries.
+	buffer  []entryWithService
+	entries []entryWithService
+
+	// Store the custom labels(syslog's structured-data) for each service
+	labels map[string]string
+
+	// connection info
+	conn          net.Conn
+	address       *url.URL
+	waitReconnect time.Duration
+	closed        bool
+	hostname      string
+	data          bytes.Buffer
+}
+
+// priorityVal calculates the syslog Priority value (PRIVAL) from the given
+// Facility and Severity values. See RFC 5424, sec 6.2.1 for details.
+func priorityVal(facility, severity int) int {
+	return facility*8 + severity
+}
+
+func fillDefaultOptions(options *ClientOptions) {
+	if options.MaxRequestEntries == 0 {
+		options.MaxRequestEntries = defaultMaxRequestEntries
+	}
+	if options.SyslogInitialBackoff == 0 {
+		options.SyslogInitialBackoff = defaultSyslogInitialBackoff
+	}
+	if options.SyslogMaxBackoff == 0 {
+		options.SyslogMaxBackoff = defaultSyslogMaxBackoff
+	}
+	if options.DialTimeout == 0 {
+		options.DialTimeout = defaultDialTimeout
+	}
+}
+
+// NewClient creates a syslog client.
+func NewClient(options *ClientOptions) (*Client, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "-"
+	}
+
+	opts := *options
+	fillDefaultOptions(&opts)
+
+	u, err := url.Parse(opts.Location)
+	if err != nil || u.Host == "" {
+		u, err = url.Parse("//" + opts.Location)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if u.Scheme != "tcp" && u.Scheme != "udp" && u.Scheme != "" {
+		return nil, fmt.Errorf("invalid syslog server address scheme %q", u.Scheme)
+	}
+	if u.Scheme == "" {
+		u.Scheme = "tcp"
+	}
+
+	c := &Client{
+		hostname: hostname,
+		address:  u,
+		options:  &opts,
+		buffer:   make([]entryWithService, 2*opts.MaxRequestEntries),
+		labels:   make(map[string]string),
+	}
+	c.entries = c.buffer[:0]
+	return c, nil
+}
+
+// SetLabels formats the given service's labels into a structured-data section
+// for a syslog message, according to RFC5424 section 6
+func (c *Client) SetLabels(serviceName string, labels map[string]string) {
+	if len(labels) == 0 {
```

Comment:
This should actually be `labels == nil`, according to the interface spec.
```suggestion
	if labels == nil {
```

Reply:
This hasn't been applied yet.